airflow/operators/latest_only_operator.py

```diff
@@ -49,38 +49,40 @@ def execute(self, context):
 
         if not left_window < now <= right_window:
             logging.info('Not latest execution, skipping downstream.')
-            downstream_task_ids = context['task'].downstream_task_ids
+            downstream_tasks = context['task'].get_flat_relatives(upstream=False)
+            downstream_task_ids = [d.task_id for d in downstream_tasks]
+
+            logging.debug("Downstream task_ids {}".format(downstream_task_ids))
+
             if downstream_task_ids:
                 session = settings.Session()
-                TI = TaskInstance
-                tis = session.query(TI).filter(
-                    TI.execution_date == context['ti'].execution_date,
-                    TI.task_id.in_(downstream_task_ids)
-                ).with_for_update().all()
-
-                for ti in tis:
-                    logging.info('Skipping task: %s', ti.task_id)
-                    ti.state = State.SKIPPED
-                    ti.start_date = now
-                    ti.end_date = now
-                    session.merge(ti)
+                if context['dag_run']:
+                    dr = context['dag_run']
+                    tis = dr.get_task_instances(session=session)
+                    for ti in tis:
+                        if ti.task_id not in downstream_task_ids:
+                            continue
 
-                # this is defensive against dag runs that are not complete
-                for task in context['task'].downstream_list:
-                    if task.task_id in tis:
-                        continue
+                        logging.info("Skipping task: {}".format(ti.task_id))
+                        ti.refresh_from_db(lock_for_update=True, session=session)
+                        ti.state = State.SKIPPED
+                        ti.start_date = now
+                        ti.end_date = now
 
-                    logging.warning("Task {} was not part of a dag run. "
-                                    "This should not happen."
-                                    .format(task))
-                    now = datetime.datetime.now()
-                    ti = TaskInstance(task, execution_date=context['ti'].execution_date)
-                    ti.state = State.SKIPPED
-                    ti.start_date = now
-                    ti.end_date = now
-                    session.merge(ti)
+                    session.commit()
```

Comment:
Should we commit once or multiple times? Otherwise we are holding a lot of locks, but it's probably fine either way